### PR TITLE
ncp functional tests

### PR DESCRIPTION
**Pull request by turon**
Added functional coverage tests for ncp firmware in tests/scripts/ncp.  These are targeting increased coverage of ncp_base and ultimately the underlying ot apis as triggered via spinel.  Over time this can be expanded to include diags and other commands not covered by thread-cert.

### CODE REVIEW

File: .travis/before_install.sh [hunk 1]
```diff
@@ -45,6 +45,11 @@ cd /tmp || die
     sudo -H pip install pexpect || die
     pip install pexpect || die
 
+    # posix-ncp tests
+    pip install scapy || die
```
Comment (jwhui) [inline .travis/before_install.sh]: NCP tests were separated from CLI tests.  This PR integrates some NCP tests with the CLI tests.  Let's keep the model consistent.

File: tests/scripts/Makefile.am [hunk 2]
```diff
@@ -28,232 +28,56 @@
...
-    thread-cert/Cert_5_5_08_SplitRoutersLostLeader.py                \
-    thread-cert/Cert_5_6_01_NetworkDataRegisterBeforeAttachLeader.py \
-    thread-cert/Cert_5_6_02_NetworkDataRegisterBeforeAttachRouter.py \
+PRETTY_SUBDIRS                          = \
     $(NULL)
 
-endif # OPENTHREAD_TESTS_SUBSET4
+if OPENTHREAD_BUILD_TESTS
+if OPENTHREAD_BUILD_COVERAGE
+CLEANFILES                             = $(wildcard *.gcda *.gcno)
```
Comment (jwhui) [inline tests/scripts/Makefile.am]: Why is this here?

File: tests/scripts/Makefile.am [hunk 3]
```diff
@@ -28,232 +28,56 @@
...
-    thread-cert/Cert_6_6_01_KeyIncrement.py                          \
-    thread-cert/Cert_6_6_02_KeyIncrementRollOver.py                  \
-    thread-cert/Cert_7_1_01_BorderRouterAsLeader.py                  \
-    thread-cert/Cert_7_1_02_BorderRouterAsRouter.py                  \
-    thread-cert/Cert_7_1_03_BorderRouterAsLeader.py                  \
-    thread-cert/Cert_7_1_04_BorderRouterAsRouter.py                  \
-    thread-cert/Cert_7_1_05_BorderRouterAsRouter.py                  \
-    thread-cert/Cert_9_2_14_PanIdQuery.py                            \
-    $(NULL)
+OPENTHREAD_COVERAGE_BUNDLE             = ${abs_builddir}/${PACKAGE}${NL_COVERAGE_BUNDLE_SUFFIX}
```
Comment (jwhui) [inline tests/scripts/Makefile.am]: Why is this here?

File: tests/scripts/ncp/Makefile.am [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/ncp/Makefile.am]: Update copyright.

File: tests/scripts/ncp/test_ncp_basics.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/python
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_basics.py]: Update copyright.

File: tests/scripts/ncp/node.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/python
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/ncp/node.py]: Update copyright.

File: tests/scripts/ncp/test_ncp_ipv6.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/python
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_ipv6.py]: Update copyright.

File: tests/scripts/ncp/test_ncp_mac.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/python
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_mac.py]: Update copyright.

File: tests/scripts/ncp/test_ncp_net.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/python
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_net.py]: Update copyright.

File: tests/scripts/ncp/test_ncp_phy.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/python
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_phy.py]: Update copyright.

File: tests/scripts/ncp/test_ncp_thread.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/python
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_thread.py]: Update copyright.

File: tests/scripts/thread-cert/Makefile.am [hunk 12]
```diff
@@ -0,0 +1,259 @@
+#
+#  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (jwhui) [inline tests/scripts/thread-cert/Makefile.am]: Update copyright.

File: tests/scripts/ncp/test_ncp_basics.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+class test_ncp(unittest.TestCase):
+    def setUp(self):
+        self.node = node.Node(1)
+
+    def test_version(self):
+        self.node.send_command('version')
+        self.node.pexpect.expect('OPENTHREAD/')
+
+    def test_scan(self):
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_basics.py]: Shouldn't this be in `test_ncp_mac`?

File: tests/scripts/ncp/test_ncp_basics.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        self.node.send_command('scan')
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_basics.py]: If this is a functional test, it would nice to actually test the scan function.

File: tests/scripts/ncp/test_ncp_basics.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        self.node.pexpect.expect('Done')
+
+    #def test_noop(self): pass
+    #def test_reset(self): pass
+
+    def test_status(self): 
+        self.node.send_command('ncp-status')
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_basics.py]: Does this actually test the `status` function?

File: tests/scripts/ncp/test_ncp_mac.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+    def test_mac_laddr(self):
+        self.node.send_command('ncp-mac64')
+        self.node.pexpect.expect('Done')
+
+    def test_mac_filter_mode(self):
+        self.node.send_command('mac-filter-mode')
+        self.node.pexpect.expect('Done')
+
+        self.node.send_command('mac-filter-mode 1')
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_mac.py]: Again, would be nice to actually test the filtering.

File: tests/scripts/ncp/test_ncp_mac.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+    def test_whitelist(self):
+        self.node.send_command('whitelist')
+        self.node.pexpect.expect('Disabled')
+        self.node.pexpect.expect('Done')
+
+        self.node.send_command('whitelist enable')
+        self.node.pexpect.expect('Done')
+
+        self.node.send_command('whitelist')
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_mac.py]: Would be nice to actually test the whitelist.

File: tests/scripts/ncp/test_ncp_thread.py [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+    def test_route(self):
+        self.node.send_command('route add fd00::1/64')
+        self.node.pexpect.expect('Done')
+
+    def test_leader_weight(self):
+        self.node.send_command('leaderweight')
+        self.node.pexpect.expect('Done')
+
+        self.node.send_command('leaderweight 4')
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_thread.py]: This will be covered by cert testing.

File: tests/scripts/ncp/test_ncp_thread.py [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+        self.node.pexpect.expect('Done')
+        self.node.send_command('ipaddr')
+        self.node.pexpect.expect('fd00::1')
+        self.node.pexpect.expect('Done')
+
+        #self.node.send_command('ipaddr remove fd00::1')
+        #self.node.pexpect.expect('Done')
+
+    def test_route(self):
+        self.node.send_command('route add fd00::1/64')
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_thread.py]: Will be covered by cert testing.

File: tests/scripts/ncp/test_ncp_mac.py [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+        self.node.pexpect.expect('Enabled')
+        self.node.pexpect.expect('Done')
+
+        self.node.send_command('whitelist add deadbeeff00dbabe')
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_mac.py]: Isn't this covered by the cert suite?

File: tests/scripts/ncp/test_ncp_net.py [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+
+import node
+
+class test_ncp_net(unittest.TestCase):
+    def setUp(self):
+        self.node = node.Node(1)
+        self.node.send_command('panid 1')
+        self.node.pexpect.expect('Done')
+
+    def test_net_name(self):
```
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_net.py]: These are thread specific, shouldn't these be in `test_ncp_thread`?

**Comment by jwhui** [comment]
I glanced through this PR.  Some high-level comments:

1. Since these are functional tests, I would prefer if they actually tested the function itself rather than just checking if spinel-cli returns `Done` or echoes the value.
2. We should remain consistent about whether the CLI and NCP tests are integrated or separated.
3. I'm surprised that some of these are not already covered by the cert tests.   For example, whitelist add/remove, etc.

**Comment by jwhui** [comment]
Also, how are you doing on #517.  Getting full support for the existing thread-cert tests is higher priority.

File: .travis/before_install.sh (hunk 1 above)
Comment (turon) [inline .travis/before_install.sh]: Agree.  The NCP tests trigger on `if OPENTHREAD_ENABLE_NCP` in `tests/scripts/ncp/Makefile.am`.  It looks like posix-distcheck and posix-32-bit configure with --enable-ncp.

File: tests/scripts/ncp/test_ncp_mac.py (hunk 20 above)
Comment (jwhui) [inline tests/scripts/ncp/test_ncp_mac.py]: Check again, `remove` is a part of the current test suite.  You are probably not seeing coverage in NCP due to #517.

**Comment by jwhui** [comment]
Thanks for attempting to address my comments.  I'd really prefer to have the code coverage results be the result of proper testing methodology.  In my view, you should never be writing test just for code coverage's sake.  The purpose of code coverage is to highlight features that have not yet been covered properly by existing scenarios and cause the tester to rethink how those features should actually be tested.
